sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,218 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: profile=True
+#
+# Author: Andreas Mueller
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+cimport numpy as np
+cimport cython
+
+from libc.math cimport sqrt
+
+from ..metrics import euclidean_distances
+from ._k_means import _centers_dense
+from ..utils.fixes import partition
+
+
+cdef double euclidian_dist(double* a, double* b, int n_features) nogil:
+    cdef double result, tmp
+    result = 0
+    cdef int i
+    for i in range(n_features):
+        tmp = (a[i] - b[i])
+        result += tmp * tmp
+    return sqrt(result)
+
+
+cdef update_labels_distances_inplace(
+        double* X, double* centers, double[:, :] center_distances, int[:] labels,
+        double[:, :] lower_bounds, double[:] upper_bounds, int n_samples, int
```

Comment:
As `center_distances` is in fact the half distances maybe you can call it `center_half_distances`. It will be easier to understand the code.